assignments/burmanm/week-5.scm

```diff
@@ -0,0 +1,583 @@
+;; 88
+;; ;Value: ()
+
+;; Exercise 2.24
+
+;; scheme@(guile-user)> (list 1 (list 2 (list 3 4)))
+;; $4 = (1 (2 (3 4)))
+
+;; "draw" :
+
+;; (1) -> (2) -> (3 4)
+
+;; Exercise 2.25
+
+;; (1 3 (5 7) 9) = (cdr (car (cdr (cdr
+
+;; scheme@(guile-user) [7]> (cdr (car (cdr (cdr (list 1 3 (cons 5 7) 9)))))
+;; $14 = 7
+
+;; ((7)) = (car (car))
+
+;; scheme@(guile-user) [8]> (car (car '((7))))
+;; $19 = 7
+
+;; (1 (2 (3 (4 (5 (6 7))))))
+
+;; scheme@(guile-user) [13]> (car (cdr (car (cdr (car (cdr (car (cdr (car (cdr (car (cdr (list 1 (list 2 (list 3 (list 4 (list 5 (list 6 7))))))))))))))))))
+;; $34 = 7
+
+;; Exercise 2.26
+
+;; scheme@(guile-user) [13]> (append x y)
+;; $35 = (1 2 3 4 5 6)
+;; scheme@(guile-user) [13]> (cons x y)
+;; $36 = ((1 2 3) 4 5 6)
+;; scheme@(guile-user) [13]> (list x y)
+;; $37 = ((1 2 3) (4 5 6))
+
+;; Exercise 2.27
+
+(define (deep-reverse list)
+  (if (null? list)
+      '()
+      (append (deep-reverse (cdr list)) (cons (reverse (car list)) '()))))
+
+;; scheme@(guile-user) [15]> (deep-reverse x)
+;; $44 = ((4 3) (2 1))
+
+;; Exercise 2.28
+
+(define (fringe t)
+  (cond ((null? t) '())
+	((not (pair? t)) (list t))
+	(else (append (fringe (car t))
+		      (fringe (cdr t))))))
+
+;; scheme@(guile-user) [25]> (fringe x)
+;; $60 = (1 2 3 4)
+;; scheme@(guile-user) [25]> (fringe (list x x))
+;; $61 = (1 2 3 4 1 2 3 4)
+;; scheme@(guile-user) [25]> 
+
+;; Exercise 2.29
+
+;; a)
+
+(define (make-mobile left right)
+  (list left right))
+
+(define (make-branch length structure)
+  (list length structure))
+
+(define (left-branch mobile)
+  (car mobile))
+
+(define (right-branch mobile)
+  (car (cdr mobile)))
+
+(define (branch-length mobile)
+  (car mobile))
+
+(define (branch-structure mobile)
+  (car (cdr mobile)))
+
+;; b)
+
+(define (total-weight mobile)
+  (define (count-branch b)
+    (if (not (pair? (branch-structure b)))
+	(branch-structure b)
+	(total-weight (branch-structure b))))
+  (+ (count-branch (left-branch mobile)) (count-branch (right-branch mobile))))
+
+;; Test..
+
+(define m1 (make-mobile (make-branch 1 1) (make-branch 1 2)))
+
+;; scheme@(guile-user) [33]> (total-weight m1)
+;; $78 = 3
+
+(define m2 (make-mobile (make-branch 1 1) (make-branch 1 (make-mobile (make-branch 1 2) (make-branch 1 1)))))
+
+;; (total-weight m2)
+;; $79 = 4
+
+;; c)
+
+;; Export count-branch
+
+(define (balanced? mobile)
+  (define (count-branch b)
+    (if (not (pair? (branch-structure b)))
+	(branch-structure b)
+	(total-weight (branch-structure b))))
+  (define (total-weight mobile)
+    (+ (count-branch (left-branch mobile)) (count-branch (right-branch mobile))))
+  (define (count-torque b)
+    (* (branch-length b) (count-branch b)))
+  (define (branch-balanced? b)
+    (if (pair? (branch-structure b))
+	(balanced? (branch-structure b))
+	#t))
+  (and (= (count-torque (left-branch mobile)) (count-torque (right-branch mobile)))
+       (branch-balanced? (left-branch mobile))
+       (branch-balanced? (right-branch mobile))))
```

Comment:
I see many people implement balanced? using a recursive balance check, but i don't think the author said each sub-branch needs to be balanced. I think what matter is that the top mobile is balanced.